File: WorksheetGenerator/Contract.ts
///<reference path="Scripts/typings/knockout/knockout.d.ts"/>

export interface IExercise {
    template: string;
}

export interface IExerciseGenerator {
    generate: () => IExercise;
}
    
export interface IExerciseGeneratorViewModel {
    name: string;
    template: string;
    generate: () => IExercise;
}

export interface ISubject {
    name: string;
    template: string;
    selectedExerciseGenerator: KnockoutObservable<IExerciseGeneratorViewModel>;
}

File: WorksheetGenerator/Contract.fs
module Contract

open FunScript
open FunScript.TypeScript

type IExercise =
    abstract member Template: string with get
    
type IExerciseGenerator =
    abstract member Name: string with get
    abstract member Template: string with get
    abstract member Generate: unit -> IExercise

type ISubject =
    abstract member Name: string with get
    abstract member Template: string with get
    abstract member ExerciseGenerators: IExerciseGenerator list with get
    abstract member SelectedExerciseGenerator: KnockoutObservable<IExerciseGenerator> with get
